### DISCUSSION TITLE
Function type narrowing error?

### DISCUSSION THREAD

**Discussion by jmg-duarte** [microsoft/pyright#2489] [Q&A] [answered]
Hello all, 

I have the following code:
```python
from typing import Type, TypeVar

T = TypeVar("T", int, bool)

def foo(v: T, t: Type[T]):
    print(t)

def bar(v: T, t: Type[T]):
    foo(v, t)
```

Pyright/Pylance report an error on the last line, on top of `t`:
```
Argument of type "Type[T@bar]" cannot be assigned to parameter "t" of type "Type[T@foo]" in function "foo"
  Type "Type[T@bar]" is incompatible with constrained type variable "T@bar"
```

From my (limited) understanding, this should work, could someone help me/explain to me why this is an error and alternatives to the previous code?

**Comment by erictraut** [comment, marked as answer]
Thanks for posting the question. Yes, this looks like a bug. It's not related to type narrowing, but it is a bug.

**Reply by jmg-duarte** [reply]
What is it then? I'd like to learn more!

**Reply by erictraut** [reply]
I'll need to investigate further before I can root cause the problem, but it's likely related to the TypeVar solving logic.

For more information about what is meant by "type narrowing", refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-narrowing).

**Reply by erictraut** [reply]
As I suspected, this was a bug in the TypeVar constraint solver — the subsystem that's responsible for determining the concrete type of a type variable based on arguments passed to a generic function. The bug was related to the way pyright tracks instantiable generic types (`type[T]`) versus type generic type instances (`T`). In your example, both `T` and `type[T]` were used in the function signature. The bug affected only the code path related to constrained type variables, which explains why this wasn't reported previously.

I've introduced a fix which will appear in the next release of pyright. Thanks again for reporting the problem.

**Reply by jmg-duarte** [reply]
Any idea when can I expect this to be in Pylance?

**Reply by erictraut** [reply]
We release a new version of pylance each week, typically on Wednesday. So, should be within the next day.

**Reply by jmg-duarte** [reply]
🙏 Thank you and your team for the amazing work!
